cmd/ocm/gcp/gcp-client-shim.go

```diff
@@ -0,0 +1,542 @@
+				return nil
+			}
+		}
+	}
+	if err != nil {
+		return errors.Wrap(err, "Failed to create IAM service account")
+	}
+	log.Printf("IAM service account %s created", serviceAccountId)
+	return nil
+}
+
+func (c *shim) createOrUpdateRoles(
+	ctx context.Context,
+	log *log.Logger,
+	roles []*cmv1.WifRole,
+) error {
+	for _, role := range roles {
+		if role.Predefined() {
+			continue
+		}
+		roleID := role.RoleId()
+		roleTitle := role.RoleId()
+		permissions := role.Permissions()
+		existingRole, err := c.getRole(ctx, c.fmtRoleResourceId(role))
+		if err != nil {
+			if gerr, ok := err.(*apierror.APIError); ok && gerr.GRPCStatus().Code() == codes.NotFound {
+				_, err = c.createRole(
+					ctx,
+					permissions,
+					roleTitle,
+					roleID,
+					roleDescription,
+					c.wifConfig.Gcp().ProjectId(),
+				)
+				if err != nil {
+					return errors.Wrap(err, fmt.Sprintf("Failed to create %s", roleID))
+				}
+				log.Printf("Role %q created", roleID)
+				continue
+			} else {
+				return errors.Wrap(err, "Failed to check if role exists")
+			}
+		}
+
+		// Undelete role if it was deleted
+		if existingRole.Deleted {
+			_, err = c.undeleteRole(ctx, c.fmtRoleResourceId(role))
+			if err != nil {
+				return errors.Wrap(err, fmt.Sprintf("Failed to undelete custom role %q", roleID))
+			}
+			existingRole.Deleted = false
+			log.Printf("Role %q undeleted", roleID)
+		}
+
+		// Update role if permissions have changed
+		if c.roleRequiresUpdate(permissions, existingRole.IncludedPermissions) {
+			existingRole.IncludedPermissions = permissions
+			_, err := c.updateRole(ctx, existingRole, c.fmtRoleResourceId(role))
+			if err != nil {
+				return errors.Wrap(err, fmt.Sprintf("Failed to update %s", roleID))
+			}
+			log.Printf("Role %q updated", roleID)
+		}
+	}
+	return nil
+}
+
+func (c *shim) roleRequiresUpdate(
+	newPermissions []string,
+	existingPermissions []string,
+) bool {
+	permissionMap := map[string]bool{}
+	for _, permission := range existingPermissions {
+		permissionMap[permission] = true
+	}
+	if len(permissionMap) != len(newPermissions) {
+		return true
+	}
+	for _, permission := range newPermissions {
+		if !permissionMap[permission] {
+			return true
+		}
+	}
+	return false
+}
+
+func (c *shim) bindRolesToServiceAccount(
+	ctx context.Context,
+	serviceAccount *cmv1.WifServiceAccount,
+) error {
+	serviceAccountId := serviceAccount.ServiceAccountId()
+	roles := serviceAccount.Roles()
+
+	// It was found that there is a window of time between when a service
+	// account creation call is made that the service account is not available
+	// in adjacent API calls. The call is therefore wrapped in retry logic to
+	// be robust to these types of synchronization issues.
+	return utils.DelayedRetry(func() error {
+		return c.bindRolesToPrincipal(
+			ctx,
+			fmt.Sprintf("serviceAccount:%s@%s.iam.gserviceaccount.com", serviceAccountId, c.wifConfig.Gcp().ProjectId()),
+			roles,
+		)
+	}, 10, 500*time.Millisecond)
```

Comment:
would there be benefit to making these consts ?

 how did we determine that is enough? 5 secs doesnt seem like much for syncing issues in GCP

Reply:
I will move these to constants. This value was found empirically: I found it takes about 2 seconds for each service account to be configured.